lldb/tools/lldb-dap/lldb-dap.cpp

```diff
@@ -910,6 +911,183 @@ void request_attach(const llvm::json::Object &request) {
+//     "properties": {
+//       "body": {
+//         "type": "object",
+//         "properties": {
+//           "breakpoints": {
+//             "type": "array",
+//             "items": {
+//               "$ref": "#/definitions/BreakpointLocation"
+//             },
+//             "description": "Sorted set of possible breakpoint locations."
+//           }
+//         },
+//         "required": [ "breakpoints" ]
+//       }
+//     },
+//     "required": [ "body" ]
+//   }]
+// },
+// "BreakpointLocation": {
+//   "type": "object",
+//   "description": "Properties of a breakpoint location returned from the
+//   `breakpointLocations` request.",
+//   "properties": {
+//     "line": {
+//       "type": "integer",
+//       "description": "Start line of breakpoint location."
+//     },
+//     "column": {
+//       "type": "integer",
+//       "description": "The start position of a breakpoint location. Position
+//       is measured in UTF-16 code units and the client capability
+//       `columnsStartAt1` determines whether it is 0- or 1-based."
+//     },
+//     "endLine": {
+//       "type": "integer",
+//       "description": "The end line of breakpoint location if the location
+//       covers a range."
+//     },
+//     "endColumn": {
+//       "type": "integer",
+//       "description": "The end position of a breakpoint location (if the
+//       location covers a range). Position is measured in UTF-16 code units and
+//       the client capability `columnsStartAt1` determines whether it is 0- or
+//       1-based."
+//     }
+//   },
+//   "required": [ "line" ]
+// },
+void request_breakpointLocations(const llvm::json::Object &request) {
+  llvm::json::Object response;
+  FillResponse(request, response);
+  auto *arguments = request.getObject("arguments");
+  auto *source = arguments->getObject("source");
+  std::string path = GetString(source, "path").str();
+  uint64_t start_line = GetUnsigned(arguments, "line", 0);
+  uint64_t start_column = GetUnsigned(arguments, "column", 0);
+  uint64_t end_line = GetUnsigned(arguments, "endLine", start_line);
+  uint64_t end_column =
+      GetUnsigned(arguments, "endColumn", std::numeric_limits<uint64_t>::max());
+
+  lldb::SBFileSpec file_spec(path.c_str(), true);
+  lldb::SBSymbolContextList compile_units =
+      g_dap.target.FindCompileUnits(file_spec);
+
+  // Find all relevant lines & columns
+  llvm::SmallVector<std::pair<uint32_t, uint32_t>, 8> locations;
+  for (uint32_t c_idx = 0, c_limit = compile_units.GetSize(); c_idx < c_limit;
+       ++c_idx) {
+    const lldb::SBCompileUnit &compile_unit =
+        compile_units.GetContextAtIndex(c_idx).GetCompileUnit();
+    if (!compile_units.IsValid())
+      continue;
+
+    // Go through the line table and find all matching lines / columns
+    for (uint32_t l_idx = 0, l_limit = compile_unit.GetNumLineEntries();
+         l_idx < l_limit; ++l_idx) {
+      lldb::SBLineEntry line_entry = compile_unit.GetLineEntryAtIndex(l_idx);
```

Comment:
I don't think this will be correct for header files. You also need to check that `line_entry.GetFileSpec()` matches the file you're searching for.

That said, I'm somewhat worried about the performance of this code. What's the largest file you've tested it on?

Doing the lookup in the lldb internal representation could be faster (at least we wouldn't need to do the filespec comparison each time, as we could translate it to a "support file index"). For that, I guess we'd need to expose a new SB function to support these kinds of queries.

Reply:
> I don't think this will be correct for header files. You also need to check that line_entry.GetFileSpec() matches the file you're searching for.

Good point, will fix

> That said, I'm somewhat worried about the performance of this code. What's the largest file you've tested it on?

re performance considerations, also see discussion in other thread https://github.com/llvm/llvm-project/pull/113787#discussion_r1819404026

Afaict, the algorithm itself has a similar complexity to setting a breakpoint. I.e., I am turning `O(n)` into a `O(2*n)` algorithm - which at least does not make things worse.

I do agree that doing this on the internal representation might be more efficient. Both thanks to the "support file index" optimization, and also because we don't need to wrap all non-qualifying results into `SBLineEntry`/`std::shared_ptr`s.

I decided against this, because to do so I would have to expose a new API as part of the `SB*` API layer, and afaik everything which I add there is essentially cast in stone / enshrined in the ABI, and I do expect changes as part of the "Go To" functionality.

Or is there some way how I can expose an API without immediately running into the shackles of ABI-compatibility?

Reply:
> What's the largest file you've tested it on?

20 lines 🙈

Reply:
> I would still encourage you to try this out on a large file, just to make sure it doesn't do something egregious. In llvm, SemaExpr.cpp is one of the largest files (but you can also find larger ones in the outside world).

I tried it on `SemaExpr.cpp`. Answering the `breakpointLocations` request takes something between 60 and 70 milliseconds, timed via `tail -f ~/lldb-dap.log | ts %.S`